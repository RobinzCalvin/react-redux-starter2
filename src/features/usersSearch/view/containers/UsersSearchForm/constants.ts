import { IUsersSearchFilters } from 'shared/types/githubSearch';
import { FormFieldNames, ISelectOption } from 'shared/types/form';
import { IUsersSearchFormFields } from '../../../namespace';

export const searchByOptions: Array<ISelectOption<IUsersSearchFilters['searchBy']>> = [
  { value: 'username-email', label: 'Username & email' },
  { value: 'login', label: 'Username' },
  { value: 'email', label: 'Email' },
  { value: 'fullname', label: 'Full name' },
];

export const searchTypeLabels: Record<IUsersSearchFilters['searchType'], string> = {
  both: 'users & organizations',
  org: 'organizations',
  user: 'users',
};

export const perPageOptions: Array<ISelectOption<IUsersSearchFilters['perPage']>> = [
  { value: 30, label: 30 },
  { value: 50, label: 50 },
  { value: 100, label: 100 },
];

export const formInitialValues: IUsersSearchFormFields = {
  searchString: '', searchBy: searchByOptions[0].value, searchType: 'both', perPage: perPageOptions[0].value,
};

export const fieldNames: FormFieldNames<IUsersSearchFormFields> = {
  searchString: 'searchString',
  searchBy: 'searchBy',
  searchType: 'searchType',
  perPage: 'perPage',
  reposLanguage: 'reposLanguage',
  maxRepos: 'maxRepos',
  minRepos: 'minRepos',
};
