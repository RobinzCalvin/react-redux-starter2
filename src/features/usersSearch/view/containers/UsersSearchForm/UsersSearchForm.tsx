import React from 'react';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';

import { IAppReduxState } from 'shared/types/app';
import { SearchForm } from 'shared/view/components';

import { selectors, actions } from './../../../redux';
import { IUsersSearchFormFields } from '../../../namespace';
import { formInitialValues, fieldNames } from './constants';
import UsersSearchSettings from './UsersSearchSettings/UsersSearchSettings';

interface IOwnProps {
  onSubmit(values: IUsersSearchFormFields): void;
}

interface IStateProps {
  isUsersSearchRequesting: boolean;
}

interface IActionProps {
  searchUser: typeof actions.searchUser;
  resetSearchResults: typeof actions.resetSearchResults;
}

type IProps = IOwnProps & IStateProps & IActionProps;

function mapDispatch(dispatch: Dispatch): IActionProps {
  return bindActionCreators({
    searchUser: actions.searchUser,
    resetSearchResults: actions.resetSearchResults,
  }, dispatch);
}

function mapState(state: IAppReduxState): IStateProps {
  return {
    isUsersSearchRequesting: selectors.selectCommunication(state, 'searchUser').isRequesting,
  };
}

function UsersSearchForm(props: IProps) {
  const { isUsersSearchRequesting } = props;
  return (
    <SearchForm<IUsersSearchFormFields>
      isSearchRequesting={isUsersSearchRequesting}
      searchInputName={fieldNames.searchString}
      onSubmit={handleFormSubmit}
      initialValues={formInitialValues}
      renderSettings={UsersSearchSettings}
      resetSearchResults={props.resetSearchResults}
    />
  );

  function handleFormSubmit(formValues: IUsersSearchFormFields) {
    const { searchUser, onSubmit } = props;
    searchUser({ searchOptions: formValues, page: 1 });
    onSubmit(formValues);
  }
}

export default connect(mapState, mapDispatch)(UsersSearchForm);
