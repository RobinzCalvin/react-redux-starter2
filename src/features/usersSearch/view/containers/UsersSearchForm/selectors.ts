import { createSelector } from 'reselect';

import { tKeys } from 'services/i18n';
import { getSelectValuesToLabelsMap } from 'shared/helpers';
import { ISelectOption } from 'shared/types/form';
import { IUsersSearchFilters } from 'shared/types/githubSearch';

import { IUsersSearchFormProps } from './UsersSearchForm';

type OptionType = ISelectOption<IUsersSearchFilters['searchBy']>;
type LabelsType = Record<IUsersSearchFilters['searchFor'], string>;

const { userSearch: intl } = tKeys.features;

export const selectFiltersLabels = createSelector(
  (props: IUsersSearchFormProps) => props.t,
  (t): Record<keyof IUsersSearchFilters, string> => ({
    searchBy: t(intl.searchBy),
    searchFor: t(intl.searchFor),
    perPage: t(intl.resultsPerPage),
    reposLanguage: t(intl.repositoriesLanguage),
    minRepos: t(intl.minRepos),
    maxRepos: t(intl.maxRepos),
  }),
);

export const selectOptions = createSelector(
  (props: IUsersSearchFormProps) => props.t,
  (t): OptionType[] => ([
    { value: 'username-email', label: t(intl.usernameAndEmail) },
    { value: 'login', label: t(intl.username) },
    { value: 'email', label: t(intl.email) },
    { value: 'fullname', label: t(intl.fullName) },
  ]),
);

export const selectLabels = createSelector(
  (props: IUsersSearchFormProps) => props.t,
  (t): LabelsType => ({
    both: t(intl.usersAndOrganizations),
    org: t(intl.organizations),
    user: t(intl.users),
  }),
);

type ValueFormatter<T> = (x: T) => any;
interface IUserSearchFiltersFormattersMap {
  searchBy: ValueFormatter<IUsersSearchFilters['searchBy']>;
  searchFor: ValueFormatter<IUsersSearchFilters['searchFor']>;
  [key: string]: unknown;
}

export const selectFiltersValuesFormatters = createSelector(
  selectOptions,
  selectLabels,
  (options, labels): IUserSearchFiltersFormattersMap => ({
    searchBy: searchByValue => getSelectValuesToLabelsMap(options)[searchByValue].toLowerCase(),
    searchFor: searchForValue => labels[searchForValue].toLowerCase(),
  }),
);
