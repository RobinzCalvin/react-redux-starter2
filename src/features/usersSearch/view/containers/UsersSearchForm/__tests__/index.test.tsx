import React from 'react';
import { shallow } from 'enzyme';

import { UsersSearchForm, IUsersSearchFormProps } from '../UsersSearchForm';

const props: IUsersSearchFormProps = {
  onSubmit: jest.fn(),
  searchUser: jest.fn(),
  resetSearchResults: jest.fn(),
  isUsersSearchRequesting: false,
};

describe('UsersSearchForm component', () => {
  const component = shallow(<UsersSearchForm {...props} />);
  it('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
