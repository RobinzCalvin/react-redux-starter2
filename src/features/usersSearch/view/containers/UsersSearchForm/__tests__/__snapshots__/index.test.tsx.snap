// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersSearchForm component should match snapshot 1`] = `
<SearchForm
  initialValues={
    Object {
      "perPage": 30,
      "searchBy": "username-email",
      "searchString": "",
      "searchType": "both",
    }
  }
  isSearchRequesting={false}
  onSubmit={[Function]}
  renderSettings={[Function]}
  resetSearchResults={[MockFunction]}
  searchInputName="searchString"
/>
`;
